File: demos/svg/svg_pad4.js
/*
 * This file defines a single function SVG to use the module
 * var x = new SVG();
 * x.mk_canvas({width:200, id:1, ht:100, color:blue});
 * x.mk_circle({parent:1, cx:10, cy:20, r:15, color:"green"});
 * x.render([cmd]) interprets a list of commands
 *  
 */

var SVG = {};

SVG.svg_ns = 'http://www.w3.org/2000/svg';
SVG.xhtml_ns = 'http://www.w3.org/1999/xhtml';
SVG.xlinkns = 'http://www.w3.org/1999/xlink';

// drag variables
SVG.xstart = 0;
SVG.ystart = 0;
SVG.dragging = false;
SVG.dragobj = null;
SVG.canvas = null;


SVG.init1 = function(o){
    // console.log('***** INIT ***** ', [SVG,o.indiv]);
    var canvas = SVG.add_canvas(o);
}
    
SVG.id = function(tag) {
    return document.getElementById(tag);
};

SVG.D = function(x, y) {
    return x == undefined ? y : x;
};

SVG.save = function(key, val){
    if(key !== undefined){
	if(V.hasOwnProperty(key)){
	    alert('Error SVG created object with duplicate ID = ' + key);
	} else {
	    V[key] = val;
	}
    }
}

SVG.add_canvas = function(o) {
    var canvas = document.createElementNS(SVG.svg_ns, 'svg');
    canvas.setAttribute("width", SVG.D(o.width, 200));
    canvas.setAttribute("height", SVG.D(o.ht, 100));
    canvas.setAttribute("style","background-color:" + 
			SVG.D(o.color, "#eeffbb"));
    canvas.setAttribute("id", o.id);
    canvas.addEventListener("mousemove", SVG.mouse_move, false);
    canvas.addEventListener("mousedown", SVG.mouse_down, false);
    canvas.addEventListener("mouseup", SVG.mouse_up, false);
    SVG.id(o.parent).appendChild(canvas);
    // add a marker that be used to add the tip to the end of the arrows
    var marker = SVG.make_arrow_marker();
    canvas.appendChild(marker);
    SVG.save(o.id, canvas);
    return canvas;
};

// Jarc_arc code from osé M. Vidal
// http://jmvidal.cse.sc.edu/talks/canvassvg/

SVG.add_arc = function(o)
{
    var path = document.createElementNS(SVG.svg_ns, "path");
    var startangle = o.startangle*Math.PI/180;
    var angle = o.angle*Math.PI/180;
    var endangle = startangle + angle;
    
    // Compute the two points where our wedge intersects the circle
    // These formulas are chosen so that an angle of 0 is at 12 o'clock
    // and positive angles increase clockwise.
    var x1 = o.cx + o.r * Math.sin(startangle);
    var y1 = o.cy - o.r * Math.cos(startangle);
    var x2 = o.cx + o.r * Math.sin(endangle);
    var y2 = o.cy - o.r * Math.cos(endangle);
    
    // This is a flag for angles larger than than a half circle
    var big = 0;
    if (endangle - startangle > Math.PI) big = 1;
    
    var d = "M " + o.cx + "," + o.cy +  // Start at circle center
    " L " + x1 + "," + y1 +             // Draw line to (x1,y1)
    " A " + o.r + "," + o.r +           // Draw an arc of radius r
    " 0 " + big + " 1 " +               // Arc details...
    x2 + "," + y2 +                     // Arc goes to to (x2,y2)
    " Z";                               // Close path back to (cx,cy)
    path.setAttribute("d", d);          // Set this path 
    path.setAttribute("fill", SVG.D(o.fill, "blue")); // Set wedge color
    path.setAttribute("stroke", "black");   // Outline wedge in black
    path.setAttribute("stroke-width", "2"); // 2 units thick
    V[o.parent].appendChild(path);
    SVG.save(o.id, path);
};

SVG.add_bezier = function(o)
{
    var path = document.createElementNS(SVG.svg_ns, "path");
    path.setAttribute("d", o.path);                         // Set this path 
    path.setAttribute("fill", SVG.D(o.fill, "blue"));       // Set wedge color
    path.setAttribute("stroke", "black");                   // Outline in black
    path.setAttribute("stroke-width", SVG.D(o.width, "2")); // 2 units thick
    V[o.parent].appendChild(path);
    SVG.save(o.id, path);
};  

SVG.add_path = function(o)
{
    var group = document.createElementNS(SVG.svg_ns, 'g');
    group.setAttribute('transform', o.transform);
    V[o.parent].appendChild(group);
    var path = document.createElementNS(SVG.svg_ns, "path");
    path.setAttribute("d", o.d);                      // Set this path 
    path.setAttribute("fill", SVG.D(o.fill, "blue")); // Set wedge color
    path.setAttribute("stroke", "black");             // Outline wedge in black
    path.setAttribute("stroke-width", SVG.D(o.width, "2")); // 2 units thick
    group.appendChild(path);
};  

SVG.add_circle = function(o)
{
    var obj = document.createElementNS(SVG.svg_ns, 'circle');
    obj.setAttribute('cx', o.x);
    obj.setAttribute('cy', o.y);
    obj.setAttribute('r', SVG.D(o.r, 10));
    obj.setAttribute('fill', SVG.D(o.fill,"red"));
    obj.setAttribute('stroke', SVG.D(o.stroke,
				      SVG.D(o.outer, "blue")));
    obj.setAttribute('stroke-width', SVG.D(o.stroke,2));
    V[o.parent].appendChild(obj);
    SVG.save(o.id, obj);
};

SVG.add_ellipse = function(o)
{
    var obj = document.createElementNS(SVG.svg_ns, "ellipse");
    obj.setAttribute("cx", o.cx);
    obj.setAttribute("cy", o.cy);
    obj.setAttribute("rx", o.rx);
    obj.setAttribute("ry", o.ry);
    obj.setAttribute("fill", SVG.D(o.fill,"black"));
    V[o.parent].appendChild(obj);
    SVG.save(o.id, obj);
};

SVG.add_group = function(o)
{
    var group = document.createElementNS(SVG.svg_ns, 'g');
    if(SVG.D(o.draggable, false)){
	group.type="draggable";
	group.id = o.id;
    };
    var x = SVG.D(o.x,10);
    var y = SVG.D(o.y,10);
    var t = "translate("+x+","+y+")";
    group.setAttribute('transform', t);
    group.setAttribute('x',x);
    group.setAttribute('y',y);
    V[o.parent].appendChild(group);
    SVG.save(o.id, group);
    return group;
};

SVG.add_image = function(o)
{
    var img = document.createElementNS(SVG.svg_ns, "image");  
    img.setAttribute("x", SVG.D(o.x,0));
    img.setAttribute("y", SVG.D(o.y,0));
    img.setAttribute("width", SVG.D(o.width, 20));
    img.setAttribute("height", SVG.D(o.ht, 20));
    img.setAttributeNS(SVG.xlinkns, "href", o.img);
    V[o.parent].appendChild(img);
    SVG.save(o.id, img);
};

SVG.add_line = function(o)
{
    // console.log('add_line',o);
    var obj= document.createElementNS(SVG.svg_ns,"line");
    obj.setAttribute("x1", o.x1);
    obj.setAttribute("y1", o.y1);
    obj.setAttribute("x2", o.x2);
    obj.setAttribute("y2", o.y2);
    obj.setAttribute("stroke","black");
    var width = SVG.D(o.width,1);
    obj.setAttribute("stroke-width",width+"px");
    obj.setAttribute("fill",SVG.D(o.fill,"black"));
    if (o.arrow == true)
	obj.setAttribute("marker-end","url(#myArrowTip)");
    obj.setAttribute("id", o.id);
    V[o.parent].appendChild(obj);
    SVG.save(o.id, obj);
};
 
SVG.add_rect = function(o)
{
    var obj = document.createElementNS(SVG.svg_ns, 'rect');
    obj.setAttribute('x', o.x);
    obj.setAttribute('y', o.y);
    obj.setAttribute('width', SVG.D(o.width,80));
    obj.setAttribute('height', SVG.D(o.ht, 20));
    obj.setAttribute('fill', SVG.D(o.color,"#aaaaaa"));
    obj.setAttribute('stroke', SVG.D(o.stroke,SVG.
				      D(o.fill, "aaaaaa")));
    obj.setAttribute('stroke-width', SVG.D(o.thickness,1));
    obj.setAttribute("rx", SVG.D(o.rx, 3));
    obj.setAttribute("ry", SVG.D(o.ry, 3));
    obj.setAttribute("id", o.id);
    if(o.dash != undefined) 
	obj.setAttribute("style", "stroke-dasharray: "+ o.dash);
    V[o.parent].appendChild(obj);
    SVG.save(o.id,obj);
    
    return obj;
};

SVG.add_text = function(o) 
{
    var text = document.createElementNS(SVG.svg_ns, "text");
    text.setAttribute("fill", SVG.D(o.fill,"black"));
    var size = SVG.D(o.size, 1);
    text.setAttribute("font-size", size+"em");
    var font = SVG.D(o.font,"Arial");
    text.setAttribute("font-family", font);
    text.setAttribute("x", SVG.D(o.x,10));
    text.setAttribute("y", SVG.D(o.y,10));
    // start middle end
    text.setAttribute("text-anchor", SVG.D(o.anchor,"start"));
    var str = SVG.D(o.str,"** missing str in text **");
    var data = document.createTextNode(o.str);
    text.setAttribute("id", o.id);
    text.appendChild(data);
    V[o.parent].appendChild(text);
    SVG.save(o.id, text);
};

SVG.mouse_up = function(evt)
{
    // can get mouse up from a click on a button
    if(SVG.dragging){
	SVG.dragging = false;
	var t = evt.target.parentNode;
	var x  = t.getAttribute("x");
	var y  = t.getAttribute("y");
	var z  = t.getAttribute("id");
	// console.log('stop dragging', [t,z]);
	// SVG.log("stop dragging x=" + x +" y=" + y);
	send_json({cmd:'stop_drag',x:x,y:y,id:z});
	if(t.ondrop){
	    t.ondrop(x, y);
	}
    }
};
    
SVG.mouse_move = function(evt)
{ 
    if (SVG.dragging) 
	{
	    var t = evt.target;
	    var x = evt.clientX + window.scrollX;
	    var y = evt.clientY + window.scrollY;
	    // Move drag element by the same amount the cursor has moved.
	    var x1 = (x-SVG.xstart);
	    var y1 = (y-SVG.ystart);
	    if (SVG.dragobj.constrain == "h"){
		// if we want to constrain the dragging we 
		// freeze y1
		y1 = SVG.dragobj.getAttribute("y");
	    };
	    if (SVG.dragobj.constrain == "v"){
		// if we want to constrain the dragging we 
		// freeze y1
		x1 = SVG.dragobj.getAttribute("x");
	    };
	    SVG.dragobj.setAttribute("x", x1);
	    SVG.dragobj.setAttribute("y", y1);
	    var t = "translate("+x1+","+y1+")";
	    SVG.dragobj.setAttribute('transform', t);
	    SVG.update_draggables(SVG.dragobj);	    
	}
};
   
SVG.update_draggables = function(x){
    // console.log('update_draggables', [x.elastic, x.id]);
    var lines= x.elastic;
    for(var i=0; i < lines.length; i++){
	var o   = V[lines[i]];
	var g1  = V[o.id1];
	var x1  = parseInt(g1.getAttribute("x"));
	var y1  = parseInt(g1.getAttribute("y"));
	var r1  = o.r1;
	var g2  = V[o.id2];
	var x2  = parseInt(g2.getAttribute("x"));
	var y2  = parseInt(g2.getAttribute("y"));
	var r2  = o.r2;
	var obj = o.obj;
	var r1  = o.r1;
	var r2  = o.r2;
	var f   = SVG.endpoints(x1,y1,r1,x2,y2,r2);
	obj.setAttribute("x1",f.x1);
	obj.setAttribute("y1",f.y1);
	obj.setAttribute("x2",f.x2);
	obj.setAttribute("y2",f.y2);
	// console.log('here', endpoints);
    }
}

SVG.mouse_down = function(evt)
{
    var t = evt.target.parentNode;
    // console.log("clicked on type=", [t,t.type]);
    if(t.type == "button"){
	// fire the button method
	t.obj.clicked();
    } else if(t.type == "draggable") {
	SVG.dragging = true;
	SVG.dragobj = t;
	var x = evt.clientX + window.scrollX;
	var y = evt.clientY + window.scrollY;
	var cxstart = t.getAttribute("x");
	var cystart = t.getAttribute("y");
	SVG.xstart = x - cxstart;
	SVG.ystart = y - cystart;
	// console.log('start drag',[t.id,t.type,t]);
	send_json({cmd:'start_drag',x:cxstart,y:cystart,id:t.id});
	// SVG.log("start dragging t.type="+t.type+
	// " x=" +cxstart+" y="+cystart);
    }
};

SVG.make_arrow_marker = function()
{
    var marker = document.createElementNS(SVG.svg_ns, "marker");
    marker.setAttribute("id", "myArrowTip"); // <-- the name in url(#...)
    marker.setAttribute("viewBox","0 0 10 10");
    marker.setAttribute("refX",1);
    marker.setAttribute("refY",5);
    marker.setAttribute("markerUnits", 8);
    marker.setAttribute("orient","auto");
    marker.setAttribute("markerWidth",8);
    marker.setAttribute("markerHeight",6);
    var path = document.createElementNS(SVG.svg_ns, "polyline");
    path.setAttribute("points","0,0 10,5 0,10 1,5");
    path.setAttribute("fill","darkblue");
    marker.appendChild(path);
    return marker;
};

SVG.log = function(x){
    SVG.id("log").innerHTML += x + "<br>";
};

SVG.add_elastic_arrow = function(o){
    // console.log('add_elastic', o);
    var g1 = V[o.id1];
    var x1 = parseInt(g1.getAttribute("x"));
    var y1 = parseInt(g1.getAttribute("y"));
    var r1 = o.r1;
    var g2 = V[o.id2];
    var x2 = parseInt(g2.getAttribute("x"));
    var y2 = parseInt(g2.getAttribute("y"));
    var r2 = o.r2;
    var f  = SVG.endpoints(x1,y1,r1,x2,y2,r2);
    SVG.add_line({parent:o.parent,id:o.id,x1:f.x1,y1:f.y1,x2:f.x2,y2:f.y2,
		  arrow:true});
    // need to tell the drag blogs that they elastic endpoints
    // make a new_id
    var newId = o.id +'e';
    SVG.save(newId, {id1:o.id1, id2:o.id2, r1:r1, r2:r2, obj:V[o.id]});
    g1.elastic.push(newId);
    g2.elastic.push(newId);
}    

SVG.endpoints = function(x1,y1,r1,x2,y2,r2)
{
   // line between two circles
    var m,a,b,x3,y3,x4,y4;
    m = 5; // marker offset
    d = Math.sqrt((x1-x2)*(x1-x2)+(y1-y2)*(y1-y2));
    b = x1;
    a = (x2 - b)/d;
    x3 = r1*a + b;
    x4 = (d-r2-m)*a + b;
    b = y1;
    a = (y2 - b)/d;
    y3 = (r1)*a + b;
    y4 = (d-r2-m)*a + b;
    return {x1:x3, y1:y3, x2:x4, y2:y4} 
}

SVG.add_drag_rect = function(o){
    var g1 = SVG.add_group({parent:o.parent, draggable:true,
			    id:o.id, x:o.x,y:o.y});
    //  console.log('g1', [g1,o.id]);
    g1.id = o.id;
    g1.constrain = 'h';
    SVG.add_rect({parent:o.id, x:0,y:0,width:o.width,color:'pink',
		  ht:o.ht, rx:2,ry:2});
};
    


SVG.add_dragblob = function(o)
{
    // red circle - black border
    var g1 = SVG.add_group({parent:o.parent, draggable:true,
			    id:o.id, x:o.x,y:o.y});
    //  console.log('g1', [g1,o.id]);
    g1.id = o.id;
    g1.elastic = new Array();
    SVG.add_circle({parent:o.id, x:0,y:0,r:o.r,stroke:2});
};
    

/* 
 * Now for some objects that are built using SVG
 * Button
 *
 */

SVG.add_button = function(o){
    new SVG.Button(o);
}

SVG.Button = function(o)
{
    var g = SVG.add_group({parent:o.parent,id:o.id,x:o.x,y:o.y});
    this.parentId = o.id;
    this.group = g;
    g.type = "button";
    g.obj = this;
    this.width = o.width;
    this.rect = SVG.add_rect({parent:o.id,
			      id:o.id+'r',
			      x:0, y:0, str:o.str, stroke:"black", 
			      rx:1, ry:1, width:o.width, ht:25, thickness:1,
			      color:"#dddddd"});
    this.text = SVG.add_text({parent:o.id,
			      id:o.id+'t',
			      x:(o.width/2),y:18, str:o.str, 
			      anchor:"middle"});
};

SVG.Button.prototype.clicked = function(){
    // r is the dotted line rectange that appears
    // this does not need an id
    var r = SVG.add_rect({parent:this.parentId,
			  x:3,y:3,color:"none",
			  stroke:"black",
			  width: this.width - 6,
			  ht:19,
			  thickness:1, dash:"1,1"});
    var oldcolor = this.rect.getAttribute('fill');
    var g = this.group;
    var f = function(){g.removeChild(r);};
    send_json({cmd:'clicked',id:this.parentId});
    setTimeout(f, 500);
};

// make a design grid -- easier to see the coordinates

SVG.add_grid = function(o){
    var width = o.width;
    var ht = o.ht;
    var step = o.step;
    var parent = o.parent;
    // draw horizontal grid
    var i = 0;
    while(i < ht){
	SVG.add_line({parent:parent, x1:0,y1:i,x2:width,y2:i});
	SVG.add_text({parent:parent, x:5,y:i+15,str:i});
	i += step;
    };
    // draw vertical grid
    i = 0;
    while(i < width){
	SVG.add_line({parent:o.parent, x1:i,y1:0,x2:i,y2:ht});
	SVG.add_text({parent:parent, x:i+5,y:15,str:i});
	i += step;
    };
}

function send_json(o){
    send(JSON.stringify(o));
}

SVG.mk_generic = function(o){
    var obj = document.createElementNS(SVG.svg_ns, o.type);
    for(key in o){
	var val = o[key];
	obj.setAttributeNS(null, key, val);
    };
    V[o.parent].appendChild(obj);
    SVG.save(o.id, obj);
}

File: demos/svg/svg_pad4.erl
-module(svg_pad4).
-compile(export_all).

start(Ws) -> 
    process_flag(trap_exit, true),
    put(free, 1),
 
    %% INIT MUST be the first command
    make_svg_object(Ws, init1, 
	    [{id, svg}, {parent, here}, 
	     {width,800}, {ht,400}, {color,'#ddebdd'}]),
    make_svg_object(Ws, add_grid,
     		    [{parent,svg}, {step,50},{width,800}, {ht,400}]),

    make_svg_object(Ws, add_drag_rect,
		    [{parent,svg},
		     {id,3456},
		     {x,300},{y,350},{width,100},{ht,25}]),
    text(Ws, 300, 390, "Hozontally constrained draggagle"),
    add_standard_objects(Ws),
    add_generic_objects(Ws),

    make_svg_object(Ws, add_button,
		    [{parent,svg},{id,23},
		     {x,200},{y,300},{width,100},{str,<<"click me">>}
		    ]),

    make_svg_object(Ws, add_button,
	    [{parent,svg},{id,223},
	     {x,50},{y,300},{width,120},{str,<<"click 12">>}
	     ]),


    %% To make an elastic line we make two drag blobs

    make_svg_object(Ws, add_dragblob,
		    [{id,100},{parent,svg},{x,20},{y,350},
		     {r,10}]),
    
    make_svg_object(Ws, add_dragblob,
		    [{id,101},{parent,svg},{x,100},{y,350},
		     {r,10}]),

    make_svg_object(Ws, add_dragblob,
		    [{id,103},{parent,svg},{x,20},{y,250},
		     {r,10}]),
    
    make_svg_object(Ws, add_elastic_arrow, 
		    [{parent,svg},{id,268}, 
		     {id1,100},
		     {r1,10},{r2,10},
		     {id2,101}]), 
    make_svg_object(Ws, add_elastic_arrow, 
		    [{parent,svg},{id,269}, 
		     {id1,101},
		     {r1,10},{r2,10},
		     {id2,103}]), 
    make_dashed_container(Ws, rect_container()),
    loop(Ws, true).

add_generic_objects(Ws) ->
    Scale = 0.03,
    Y=230,
    make_generic_object(Ws, rect,	 
			[{x,50},{y,50},{rx,5},{ry,5},
			 {parent,svg},
			 {width,50},{height,20},
			 {fill,<<"#bacdff">>}
			]),
   
    text(Ws, 55, 65,"rect1"),
   
    make_generic_object(Ws,rect,
			[{x,50},
			 {y,100},
			 {rx,5},
			 {ry,5},{width,50},{height,30},
			 {'stroke-dasharray', <<"2,2">>},
			 {'stroke', blue},
			 {'stroke-width',3},
			 {fill,yellow},
			 {parent,svg}]),
    text(Ws, 55,120,"rect2"),   
 
    make_generic_object(Ws,circle,
	  [{r,35},
	   {cx,150},{cy, 100},
	   {'stroke-dasharray', <<"2,2">>},
	   {'stroke', blue},
	   {'stroke-width',3},
	   {fill,orange},
	   {parent,svg}]),
    text(Ws, 120,110,"circle1"),
 
    make_generic_object(Ws, circle,
			[{r,35},
			 {cx,150},{cy, 200},
			 {fill,green},
			 {parent,svg}]),
    
    text(Ws, 130,210,"circle2"),
    
    make_generic_object(Ws, path,
			[{parent,svg},
			 transform(250,100,Scale),
			 {d,wow()}]),
    text(Ws, 250, 140, "Music!"),

    %% Here X=100 Y=270
    make_generic_object(Ws, path,
			[
			 {parent,svg},
			 {style,<<"stroke:black">>},
			 %% H is horizional line to
			 %% h is in relative units (easier)
			 {d,<<"M100 280 H 250 M 100 290 h 150">>}]
		       ),
      text(Ws, 100, 270, "horizontal lines").
    

rect_container() ->
    {550,50,200,300}.


make_dashed_container(Ws, {X, Y, W, H}) ->
    make_svg_object(Ws, add_rect,[{x,X},{y,Y},{ht,H},{width,W},
			  {parent,svg},
			  {color,none},
			  {thickness,6}, 
			  {dash,<<"4,4">>}]).

add_standard_objects(Ws) ->
    make_svg_object(Ws, add_rect,[{x,200},{y,50},{ht,200},
				  {width,150},
				  {parent,svg},
				  {color,none},
				  {thickness,4}, 
				  {dash,<<"4,4">>}]),
    
    make_svg_object(Ws, add_line,[{x1,350},{y1,350},{x2,475},{y2,325},
				  {width,2},
			  {parent,svg}]),

    make_svg_object(Ws, add_line,[{parent,svg},
				  {x1,350},{y1,340},{x2,475},{y2,315},
				  {width,2}, {arrow,true}]),

    make_svg_object(Ws, add_image,[{img,<<"./p5.jpg">>}, 
				   {x,200}, 
				   {y,350}, {width,100}, {ht,50},
				   {parent,svg}]),

    make_svg_object(Ws, add_text,
		    [{parent,svg},{x,350},
		     {y,250}, {str,<<"Drag the Red blob below">>}]),

    make_svg_object(Ws, add_dragblob,[{id,12}, {parent,svg},{x,450},{y,280}]),
    
    
    %% green ellipse
    make_svg_object(Ws, add_ellipse, [{cx,300},{cy,200},{rx,45},{ry,25},
				      {fill,pink},{parent,svg}]),
    text(Ws,270,200,"ellipse1"),
    
    %%  an arcs

    make_svg_object(Ws, add_arc,[{parent,svg},
				 {fill,orange},
				 {startangle,20}, {angle,65}, 
				 {cx,300}, {cy,100}, {r,45}]), 
    text(Ws,315,75,"arc"),
    
    %% groups can be made draggable

    make_svg_object(Ws, add_group,
		    [{parent,svg},{id,99},{x,400},
		     {draggable,true},
		     {y,50}]),
    make_svg_object(Ws, add_rect,[{parent,99}, {x,0},{y,0}, 
				  {width,100},{ht,150},
				  {color,<<"#fedcba">>}]),
    make_svg_object(Ws, add_text,
		    [{parent,99},{y,20},{str,<<"Drag Me">>}]),

    make_svg_object(Ws, add_arc,[{parent,99}, {startangle,10}, {angle,320}, 
	      {cx,50}, {cy,60}, {r,35}, {fill, red}]).
    
make_generic_object(Ws, Type, L) ->
    L1 = [{cmd,'SVG.mk_generic'},{type,Type}|L],
    Ws ! {send, list_to_binary(ezwebframe_mochijson2:encode([{struct,L1}]))}.

make_svg_object(Ws, X, L) ->
    Cmd= list_to_binary(atom_to_list(X)),
    L1 = [{cmd,<<"SVG.",Cmd/binary>>}|L],
    Ws ! {send, list_to_binary(ezwebframe_mochijson2:encode([{struct,L1}]))}.


new_index() ->
    N = get(free),
    put(free, N+1),
    N.

enc(L) ->
    L1 = [{struct,J} || J <- L],
    C = ezwebframe_mochijson2:encode(L),
    list_to_binary(C).

transform(X, Y, Scale) ->
    {transform,f2b("translate(~p,~p) scale(~p,~p)", [X,Y,Scale,-Scale])}.

text(Ws, X, Y, Str) ->
    make_svg_object(Ws, add_text, 
		    [{parent,svg},{x,X},{y,Y},{str,list_to_binary(Str)}]).


random_rect(Ws) ->
    {X0,Y0,Width,Ht} = rect_container(),
    Id = new_index(),
    X = X0 + pos_ran(Width),
    Y = Y0 + pos_ran(Ht),
    W = pos_ran(Width+X0-X) - 5,
    H = pos_ran(Ht+Y0 -Y) - 5,
    if
	H < 0; W < 0 ->
	    random_rect(Ws);
	true ->
	    make_generic_object(Ws,rect,
				[{rx,3},
				 {ry,3},
				 {parent,svg},
				 {x,X},{y,Y},{width,W},{height,H},
				 {fill,ran_color()}
				])
    end.

pos_ran(N) when N =< 0 ->
    1;
pos_ran(N) ->
    random:uniform(N).

ran_color() ->
    B1 = unsigned_byte_to_hex_string(random:uniform(255)),
    B2 = unsigned_byte_to_hex_string(random:uniform(255)),
    B3 = unsigned_byte_to_hex_string(random:uniform(255)),
    list_to_binary([$#,B1,B2,B3]).

wow() ->
    <<"M643 2c0 -102 -65 -214 -190 -248c0 -13 1 -27 1 -40c0 -46 -1 -92 -4 -138c-7 -119 -92 -227 -214 -227c-111 0 -202 92 -202 205c0 58 54 104 113 104c54 0 95 -48 95 -104c0 -52 -43 -95 -95 -95c-13 0 -27 4 -39 10c26 -47 74 -80 130 -80c100 0 166 94 172 193c3 44 4 89 4 133v31c-31 -5 -63 -6 -79 -6c-189 0 -333 173 -333 372c0 181 134 314 254 451c-37 129 -54 211 -54 379c0 197 147 308 159 308c25 0 151 -219 151 -388c0 -150 -90 -267 -190 -380c22 -73 42 -147 61 -221h6c154 0 254 -127 254 -259zM452 -207c66 20 124 84 124 170c0 90 -64 178 -168 192c27 -129 40 -239 44 -362zM338 -220c7 0 45 1 75 5c-4 127 -19 241 -47 372c-87 -5 -136 -62 -136 -124c0 -45 26 -92 83 -125c4 -4 7 -9 7 -14c0 -11 -9 -21 -20 -21c-15 0 -125 63 -125 186c0 90 61 178 168 198c-16 64 -35 127 -53 190c-110 -124 -220 -249 -220 -414c0 -149 144 -253 268 -253zM409 1108c-100 -55 -162 -159 -162 -273c0 -94 27 -190 40 -236c86 102 158 209 158 342c0 77 -11 110 -36 167z">>.
    

f2b(F, D) ->
    list_to_binary(lists:flatten(io_lib:format(F, D))).


loop(Ws, B) ->
    receive
	{msg,Json} ->
	    io:format("Received:~p~n",[Json]),
	    loop(Ws, B);
	X ->
	    io:format("handler got unexpected::~p~n",[X]),
	    loop(Ws, B)
    after 1000 ->
	    case B of
		true ->
		    random_rect(Ws),
		    loop(Ws, B);
		false ->
		    loop(Ws, B)
	    end
    end.

-spec unsigned_byte_to_hex_string(integer()) -> [byte()].

%% N is in -128 .. 127    

unsigned_byte_to_hex_string(N) when N >= 0, N < 256 ->
    [nibble_to_hex_char(N bsr 4),nibble_to_hex_char(N band 15)].

nibble_to_hex_char(X) when X < 10 -> $0 + X; 
nibble_to_hex_char(X) -> $a + X - 10.
